include/eosio/vm/host_function.hpp

```diff
@@ -17,522 +21,371 @@
-   struct aligned_ref_wrapper {
-      constexpr aligned_ref_wrapper(void* ptr) : _impl(ptr) {}
-      constexpr operator T&() const {
-         return *static_cast<T*>(_impl);
-      }
+   template <typename Host, typename Execution_Interface=execution_interface>
+   struct type_converter : public running_context<Host, Execution_Interface> {
+      using base_type = running_context<Host, Execution_Interface>;
+      using base_type::running_context;
+      using base_type::get_host;
 
-      aligned_ptr_wrapper<T, Align> _impl;
-   };
+      // TODO clean this up and figure out a more elegant way to get this for the macro
+      using elem_type = operand_stack_elem;
 
-   template<typename T, std::size_t Align = alignof(T)>
-   struct aligned_ptr {
-      aligned_ptr(const aligned_ptr_wrapper<T, Align>& wrapper) : _ptr(wrapper) {}
-      operator T*() const { return _ptr; }
-      T* _ptr;
-   };
+      EOS_VM_FROM_WASM(bool, (uint32_t value)) { return value ? 1 : 0; }
+      uint32_t to_wasm(bool&& value) { return value ? 1 : 0; }
+      template<typename T>
+      no_match_t to_wasm(T&&);
 
-   template<typename T, std::size_t Align = alignof(T)>
-   struct aligned_ref {
-      aligned_ref(const aligned_ptr_wrapper<T, Align>& wrapper) : _ptr(wrapper) {}
-      operator T&() const { return *_ptr; }
-      T* _ptr;
-   };
+      template <typename T>
+      auto from_wasm(void* ptr, wasm_size_t len) const
+         -> std::enable_if_t<is_span_type_v<T>, T> {
+         this->template validate_pointer<typename T::value_type>(ptr, len);
+         return {static_cast<typename T::pointer>(ptr), len};
+      }
 
-   // This class can be specialized to define a conversion to/from wasm.
-   template <typename T>
-   struct wasm_type_converter {};
+      template <typename T>
+      auto from_wasm(void* ptr, wasm_size_t len) const
+         -> std::enable_if_t< is_argument_proxy_type_v<T> &&
+                              is_argument_proxy_legacy_v<T> &&
+                              is_span_type_v<dependent_type_t<T>>, T> {
+         this->template validate_pointer<argument_proxy_dependent_type_t<T>>(ptr, len);
+         return {ptr, len};
+      }
+
+      template <typename T>
+      auto from_wasm(void* ptr) const
+         -> std::enable_if_t< is_argument_proxy_type_v<T> &&
+                              is_argument_proxy_legacy_v<T> &&
+                              !is_span_type_v<dependent_type_t<T>>, T> {
+         this->template validate_pointer<argument_proxy_dependent_type_t<T>>(ptr, 1);
+         return {ptr};
+      }
 
-   struct linear_memory_access {
-      void* get_ptr(uint32_t val) const {
-         return _alloc->template get_base_ptr<char>() + val;
+      template <typename T>
+      auto from_wasm(void* ptr) const
+         -> std::enable_if_t< is_argument_proxy_type_v<T> &&
+                              !is_argument_proxy_legacy_v<T> &&
+                              !is_span_type_v<dependent_type_t<T>>, T> {
```

Comment:
This `from_wasm` and the one above could just be combined into one that checks `is_argument_proxy_type_v<T> && std::is_pointer_v<typename T::proxy_type>`.